### PR TITLE
Feature/csharp8

### PR DESCRIPTION
**Pull request by david-driscoll**
Builds on #164 

This wip initial progress for C# 8 features.

* [ ] Readonly members
* [ ] Default interface members
* [x] Pattern matching enhancements:
  * [x] Switch expressions
  * [x] Property patterns
  * [x] Tuple patterns
  * [x] Positional patterns
* [ ] Using declarations
* [ ] Static local functions
* [ ] Disposable ref structs
* [ ] Nullable reference types
* [x] Asynchronous streams
* [ ] Indices and ranges
* [ ] Null-coalescing assignment
* [ ] Unmanaged constructed types
* [ ] stackalloc in nested expressions
* [x] Enhancement of interpolated verbatim strings

### CODE REVIEW

**Comment by say25** [comment]
Any updates on this? Is it possible to merge down features as they are implemented?

It seems that GitHub doesn't currently support syntax highlighting for C#8 features due to this.

See below example:
https://github.com/PostgreSQLCopyHelper/PostgreSQLCopyHelper/blob/5faeec8e5de1c81a97433e7dd2e7774006f1d128/src/PostgreSQLCopyHelper/PostgreSQLCopyHelper.cs#L129-L163

**Comment by david-driscoll** [comment]
ping @DustinCampbell @JoeRobich

**Comment by am11** [comment]
👍 
I was wondering instead of the handwritten ones, if it is it achievable to write a tool which can emit the tmLangauge patterns by parsing using: https://github.com/dotnet/roslyn/blob/version-3.2.0/src/Compilers/CSharp/Portable/Syntax/Syntax.xml? That can be run once per language version release, as `Syntax.xml` is used by compiler itself and some other tools (including ANTLR4 grammar writer) in Roslyn repository. more at: https://github.com/dotnet/csharp-tmLanguage/issues/168

**Comment by JoeRobich** [comment]
@david-driscoll Sorry for the delay in reviewing. I wasn't sure if you were trying to check all the boxes in this PR or not. Will take a look at it today.

**Comment by JoeRobich** [comment]
@david-driscoll I pulled and ran the tests locally. The tests are failing for me because the `InterpolatedString.VerbatimBeginReverse` const is missing. Also, the expression test 'works with property pattern' needs a few fix-ups to pass. I am going to have to come back to this when I have a bit more time to grok the grammar changes.

File: test/interpolated-string.tests.ts [hunk 1]
```diff
@@ -148,6 +148,29 @@ world!";`);
                 Token.Punctuation.Semicolon]);
         });
 
+        it("verbatim with two interpolations (reverse)", async () => {
```
Comment (NTaylorMullen) [inline test/interpolated-string.tests.ts]: You guys might want to consider moving to a snapshot model for grammar testing. I started doing this for our [Razor grammar](https://github.com/aspnet/AspNetCore-Tooling/tree/master/src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.Grammar.Test/tests) and it's made testing super easy, non-manual and ensures we don't break anything.

- [Example](https://github.com/aspnet/AspNetCore-Tooling/blob/2fae94440496d57b69ddbc93e038b3509de1c931/src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.Grammar.Test/tests/CodeBlock.ts#L47-L52) of a test
- [Example ](https://github.com/aspnet/AspNetCore-Tooling/blob/2fae94440496d57b69ddbc93e038b3509de1c931/src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.Grammar.Test/tests/__snapshots__/GrammarTests.test.ts.snap#L3788-L3812) of the snapshot it generates / compares

We're using the [jest snapshot testing model](https://jestjs.io/docs/en/snapshot-testing) which takes care of updating/comparing the snapshots that we generate [with our custom factory](https://github.com/aspnet/AspNetCore-Tooling/blob/2fae94440496d57b69ddbc93e038b3509de1c931/src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.Grammar.Test/tests/infrastructure/SnapshotFactory.ts).

**Pull request merged**